DSharpPlus/Entities/Interaction/DiscordInteraction.cs

```diff
@@ -168,7 +168,7 @@ public DiscordChannel Channel
     /// </summary>
     /// <param name="type">The type of the response.</param>
     /// <param name="builder">The data, if any, to send.</param>
-    public virtual async Task CreateResponseAsync(DiscordInteractionResponseType type, DiscordInteractionResponseBuilder builder = null)
+    public virtual async Task<DiscordMessage> CreateResponseAsync(DiscordInteractionResponseType type, DiscordInteractionResponseBuilder builder = null)

```

Comment:
Would it make sense to not expose `CreateResponseAsync` with a return type of `Task<DiscordMessage>`, but instead abstract this to something like `DiscordApiInteraction`?

This would eliminate the implicit API call in the `DiscordHttpInteraction` class, and keep the exposure of `DiscordMessage` only in the case when responding with the api, and (generally) would be backwards compatible with the existing codebase prior to this change since the new behaviour would only be expose on a new type.

The hierarchy could look like the following:
```
              ┌──────────────────┐               
              │DiscordInteraction│               
              └────────┬─────────┘               
                       │                         
          ┌────────────┴────────────┐            
          │                         │            
┌─────────┴───────────┐  ┌──────────┴───────────┐
│DiscordApiInteraction│  │DiscordHttpInteraction│
└─────────────────────┘  └──────────────────────┘
```

`DiscordApiInteraction` would override `CreateResponseAsync` with the return type of `Task<DiscordMessage>`, and `DiscordInteraction` would become completely abstract.

Reply:
Having different behavior between "normal" interactions and HTTP interactions was something I considered, but didn't like the prospect of having to potentially cast to a different type to get this behavior, especially considering that for the case of HttpInteractions, this would lead to even more duplicated for responses - the most pressing area this affects is commands. Another idea I had was to have the methods return `DiscordMessage?`, as to absolve the extra request for HTTP interactions, but that also didn't seem like very good DX. cc @akiraveliara, and @Plerx2493

Reply:
I also agree that there can be tradeoffs with abstraction here, but imo the overhead of forcing http interactions to preform a non-configurable implicit API request is horrible.

The selling point of HTTP interactions is you detach the reliance of the rest api, discord provides enough data per interaction to allow you to process it gracefully. When you force an API request (even if the user doesn't end up needing/caring about the message) you're adding a cost, which contributes to the ratelimit and execution time of responding.

Ideally there should be a solution to where if the user needed the response message, that its available, otherwise there shouldn't be an attempt to provide it. One other approach to this would be to define a second response function that returns a `Task<DiscordMessage>`. For HTTP interactions, that method would preform the get request, and for API interactions it would specify the `with_response=true` flag.

Reply:
> The selling point of HTTP interactions is you detach the reliance of the rest api,

The biggest selling point is detaching bots from the *gateway* - REST calls may still be required (e.g., followups)

> When you force an API request (even if the user doesn't end up needing/caring about the message) you're adding a cost, 

I agree, and making an API call isn't something I'm particularly fond of either

> which contributes to the ratelimit and execution time of responding.

Yesn't - I've prodded about interactions and the response was "they exist, but are very high and shouldn't matter to 99% of bots" (that is, [according to ckohen - feel free to inquire if he has an original link about it](https://canary.discord.com/channels/613425648685547541/801246036906606622/1241173759797366785) )

Nevertheless, as stated above, I don't like the API call either. As for the execution time of responding, this API call comes *after* the TCS is set, so no delay should be incurred in terms of Discord <-> App (though, this would technically slow down, say, *editing* after deferring, for example. Very unideal.)

> One other approach to this would be to define a second response function that returns a `Task<DiscordMessage>`

Actually I've just remembered that [`[NotNullWhen(true)]`](https://learn.microsoft.com/en-us/dotnet/api/system.diagnostics.codeanalysis.notnullwhenattribute?view=net-8.0) exists - I think that the solution here *is* to return `DiscordMessage?` and simply add a new boolean attributed with `[NotNullWhen]` to satisfy the compiler - additional documentation would have to be written surrounding this, but this does prevent having to special-case HTTP interactions for the most part, I think?

Reply:
> but imo the overhead of forcing http interactions to preform a non-configurable implicit API request is horrible.

100%.

guaranteeing a `DiscordMessage` return also doesn't work for unrelated reasons - CreateResponseAsync doesn't need to have anything to do with a message. how do you plan on getting the message to return if we're creating a modal or (very common case) deferring, for example?

Reply:
`NotNullWhen` is applied to parameters?

> As for the execution time of responding, this API call comes after the TCS is set, so no delay should be incurred in terms of Discord <-> App (though, this would technically slow down, say, editing after deferring, for example. Very unideal.)

TCS is good, I'm more referring to:
```cs
// my handler does some stuff

await CreateResponseAsync(...);

// my handler does some more stuff after responding
```
which adds to the cost of continuation, by having to fetch the response, even if its unused.

on a more general note, *implying* that something can have a return result (by having `Task<T>`) when in actuality the method always returns null for http interactions is also bad DX. This also converts a what would be cast check to a null check (or even just a cast check again). Consider the following with the proposed API:

```cs
public async Task MyHandler(CommandContext context)
{
    ...

    var message = await context.RespondAsync(...);

    // to even verify we can work with the message, we need to always null check it now. (or assert not null in cases where we as the developer infer that it cannot be null)

    if(message is not null)
    {
        // do more logic
    }
}
```

I get that via inferred context (ex: I write a non-http interaction bot) we can preform these assertions for not-null result of `message`, but that in practice may not even be the case.

*even on top of this* you're trying to provide *all* functionality under `CommandContext` which borders on the fundamentals of it being polymorphic imo.

I don't know DSharpPlus's design principles or what is classified as reasonable and consistent APIs for the library, but generally its always a good idea that the behaviour is intuitive or at least strongly documented, for whatever approach you decide to go with